cfpb/cfgov-refresh
# HOLD: Refactor context formation for pages

# Holding in favor of breaking this down
## Continuation of #2301
This refactors the way that context is set for pages. This is a rather large PR since it was combined with the above linked PR.

## The benefits of this PR:
- Decouple form logic with page type. We no longer need to create a separate page just because we want to include a form on it.
- Abstracts code away from CFGOVPage.
- Utilizes wagtail's hook functionality for inclusion
- Sets the stage for other modules coming

## What this does:
- moves "filterable" and js file context logic from util module into a handlers module
- handlers are registered in `CFGOVPage.get_context`, which is called for every page load, via `wagtail_hooks.py`
- refactors both the filterable and page js logic a bit
- provides wider test coverage


## Additions

- More tests!
- `wagtail_hooks.py` registered functions for `cfgov_context_handlers` hook to manipulate page context
- handlers module to house them
- `QueryFormMixin` to abstract the weird query stuff we're doing in the filterable form classes

## Removals

- test files where the tests are now handled elsewhere (nothing was removed that wasn't replaced)

## Changes

- `FilterableListMixin` into a handler class
- `FilterControls` is now `FilterableListControls` 
- `has_active_filters` moved to util module
- all context logic for filterable and JS handling moved from the page class to handlers
- `FilterableListControls` moved `page_type` from within `categories` StructBlock to at the base level.
- `NumberBlock` was refactored a bit for filter controls module

## Testing

- `cfgov/manage.py migrate`
- `tox`
- Make sure you go to the blog, newsroom, activity log, events archive, any browse filterable, and any sublanding filterable page. Test various filtering methods to make sure they all still work.
- Make sure JS handling still works by visiting or creating a page that has expandables on it. The expandable should work.

## Review

- @cfpb/cfgov-backends 
- @Scotchester 

## Notes

- This work is necessary to build the components necessary for Design Ops
- The Data & Research (Dada) team will be working off of a branch with these commits in order to move forward.

## Todos

- Since we no longer need them, we should remove the [Sublanding/Browse]FilterablePage classes, along with their subclasses. I felt that it would be a bit too extreme a change to include in the PR. Let me know if we should just include it.

## Checklist
